python/pyspark/sql/streaming.py

```diff
@@ -843,6 +844,169 @@ def trigger(self, processingTime=None, once=None, continuous=None):
+        ...         return True
+        ...     def process(self, row):
+        ...         print(row)
+        ...     def close(self, error):
+        ...         print("Closed with error: %s" % str(error))
+        ...
+        >>> writer = sdf.writeStream.foreach(RowPrinter())
+        """
+
+        from pyspark.rdd import _wrap_function
+        from pyspark.serializers import PickleSerializer, AutoBatchedSerializer
+        from pyspark.taskcontext import TaskContext
+
+        if callable(f):
+            """
+            The provided object is a callable function that is supposed to be called on each row.
+            Construct a function that takes an iterator and calls the provided function on each row.
+            """
+            def func_without_process(_, iterator):
+                for x in iterator:
+                    f(x)
+                return iter([])
+
+            func = func_without_process
+
+        else:
+            """
+            The provided object is not a callable function. Then it is expected to have a
+            'process(row)' method, and optional 'open(partition_id, epoch_id)' and
+            'close(error)' methods.
+            """
+
+            if not hasattr(f, 'process'):
+                raise Exception(
+                    "Provided object is neither callable nor does it have a 'process' method")
+
+            if not callable(getattr(f, 'process')):
+                raise Exception("Attribute 'process' in provided object is not callable")
+
+            open_exists = False
+            if hasattr(f, 'open'):
+                if not callable(getattr(f, 'open')):
+                    raise Exception("Attribute 'open' in provided object is not callable")
+                else:
+                    open_exists = True
+
+            close_exists = False
+            if hasattr(f, "close"):
+                if not callable(getattr(f, 'close')):
+                    raise Exception("Attribute 'close' in provided object is not callable")
+                else:
+                    close_exists = True
+
+            def func_with_open_process_close(partition_id, iterator):
+                epoch_id = TaskContext.get().getLocalProperty('streaming.sql.batchId')
+                if epoch_id:
+                    epoch_id = int(epoch_id)
+                else:
+                    raise Exception("Could not get batch id from TaskContext")
+
+                should_process = True
+                if open_exists:
+                    should_process = f.open(partition_id, epoch_id)
+
+                def call_close_if_needed(error):
+                    if open_exists and close_exists:
+                        f.close(error)
+                try:
+                    if should_process:
+                        for x in iterator:
+                            f.process(x)
+                except Exception as ex:
+                    call_close_if_needed(ex)
```

Comment:
Should it be put in finally?

Reply:
`putting this to finally` sounds like better pattern for me too, and then we may be able to inline `call_close_if_needed`.

Reply:
That is tricky because I have to pass the exception and rethrow the error if any. alternative code is
```
err = None
...
except Exception as ex:
  err = ex
finally:
  call_close(ex)
  if ex is not None: 
     raise ex
```

Reply:
Ah, sure.

Reply:
Either is fine for me.